File: java/src/nz/sodium/Node.java
package nz.sodium;

import java.util.ArrayList;
import java.util.List;
import java.util.HashSet;
import java.util.Set;
import java.lang.ref.WeakReference;

class Node implements Comparable<Node> {
    public final static Node NULL = new Node(Long.MAX_VALUE);

	Node(long rank) {
		this.rank = rank;
	}

	public static class Target {
	    Target(TransactionHandler<Unit> action, Node node) {
	        this.action = new WeakReference<TransactionHandler<Unit>>(action);
	        this.node = node;
	    }
	    final WeakReference<TransactionHandler<Unit>> action;
	    final Node node;
    }

	private long rank;
    List<Target> listeners = new ArrayList<Target>();

	/**
	 * @return true if any changes were made. 
	 */
	boolean linkTo(TransactionHandler<Unit> action, Node target, Target[] outTarget) {
		boolean changed = target.ensureBiggerThan(rank, new HashSet<Node>());
		Target t = new Target(action, target);
		listeners.add(t);
		outTarget[0] = t;
		return changed;
	}

	void unlinkTo(Target target) {
	    listeners.remove(target);
	}

	private boolean ensureBiggerThan(long limit, Set<Node> visited) {
		if (rank > limit || visited.contains(this))
			return false;

		visited.add(this);
		rank = limit + 1;
		for (Target l : listeners)
			l.node.ensureBiggerThan(rank, visited);
		return true;
	}

	@Override
	public int compareTo(Node o) {
		if (rank < o.rank) return -1;
		if (rank > o.rank) return 1;
		return 0;
	}
}

File: python/sodiumfrp/node.py
from typing import Callable, List, Set
import weakref

from sodiumfrp.transaction import Transaction
from sodiumfrp.unit import Unit

TransactionHandler = Callable[[Transaction, Unit], None]


class Target:

    def __init__(self, action: TransactionHandler, node: "Node") -> None:
        self.action = weakref.ref(action)
        self.node = node


class Node:

    def __init__(self, rank: int) -> None:
        self._rank = rank
        self._listeners: List[Target] = []


    def _link_to(self,
            action: TransactionHandler,
            target: "Node",
            out_target: List[Target]) -> bool:
        """ Return true if any changes were made. """
        changed = target._ensure_bigger_than(self._rank, set())
        t = Target(action, target)
        self._listeners.append(t)
        out_target[0] = t
        return changed


    def _unlink_to(self, target: Target) -> None:
        self._listeners.remove(target)


    def _ensure_bigger_than(self, limit: int, visited: Set["Node"]) -> bool:
        if self._rank > limit or (self in visited):
            return False

        visited.add(self)
        self._rank = limit + 1
        for listener in self._listeners:
            listener.node._ensure_bigger_than(self._rank, visited)
        visited.remove(self)
        return True


    def __lt__(self, other) -> bool:
        return self._rank < other._rank


    def __eq__(self, other) -> bool:
        return self._rank == other._rank
